quaddtype/numpy_quaddtype/src/casts.cpp

```diff
@@ -605,6 +605,175 @@ quad_to_bytes_loop(PyArrayMethod_Context *context, char *const data[],
+            if (str_descr->has_string_na || str_descr->default_string.buf != NULL) {
+                s = str_descr->default_string;
+            }
+            else {
+                NpyString_release_allocator(allocator);
+                PyErr_SetString(PyExc_ValueError, "Cannot convert null string to QuadPrecision");
+                return -1;
+            }
+        }
+
+        quad_value out_val;
+        if (bytes_to_quad_convert(s.buf, s.size, backend, &out_val) < 0) {
+            NpyString_release_allocator(allocator);
+            return -1;
+        }
+
+        store_quad<Aligned>(out_ptr, out_val, backend);
+
+        in_ptr += in_stride;
+        out_ptr += out_stride;
+    }
+
+    NpyString_release_allocator(allocator);
+    return 0;
+}
+
+// QuadDType to StringDType casting
+static NPY_CASTING
+quad_to_stringdtype_resolve_descriptors(PyObject *NPY_UNUSED(self), PyArray_DTypeMeta *dtypes[2],
+                                        PyArray_Descr *given_descrs[2], PyArray_Descr *loop_descrs[2],
+                                        npy_intp *view_offset)
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    if (given_descrs[1] == NULL) {
+        // Default StringDType() already has coerce=True
+        loop_descrs[1] = (PyArray_Descr *)PyObject_CallNoArgs(
+                (PyObject *)&PyArray_StringDType);
+        if (loop_descrs[1] == NULL) {
+            Py_DECREF(loop_descrs[0]);
+            return (NPY_CASTING)-1;
+        }
+    }
+    else {
+        Py_INCREF(given_descrs[1]);
+        loop_descrs[1] = given_descrs[1];
+    }
+
+    return NPY_SAFE_CASTING;
+}
+
+// Note: StringDType elements are always aligned, so Aligned template parameter
+// is kept for API consistency but both versions use the same logic
+template <bool Aligned>
+static int
+quad_to_stringdtype_strided_loop(PyArrayMethod_Context *context, char *const data[],
+                                 npy_intp const dimensions[], npy_intp const strides[],
+                                 void *NPY_UNUSED(auxdata))
+{
+    npy_intp N = dimensions[0];
+    char *in_ptr = data[0];
+    char *out_ptr = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+
+    PyArray_Descr *const *descrs = context->descriptors;
+    QuadPrecDTypeObject *descr_in = (QuadPrecDTypeObject *)descrs[0];
+    PyArray_StringDTypeObject *str_descr = (PyArray_StringDTypeObject *)descrs[1];
+    QuadBackendType backend = descr_in->backend;
+
+    npy_string_allocator *allocator = NpyString_acquire_allocator(str_descr);
+
+    while (N--) {
+        quad_value in_val = load_quad<Aligned>(in_ptr, backend);
+        Sleef_quad sleef_val = quad_to_sleef_quad(&in_val, backend);
+
+        // Get string representation with adaptive notation
+        // Use a large buffer size to allow for full precision
+        PyObject *py_str = quad_to_string_adaptive(&sleef_val, QUAD_STR_WIDTH);
```

Comment:
There's no need to create a PyUnicode object here. Just pass the ASCII bytes of the C parsed string to `NpyString_Pack`.

Reply:
So actually this function `quad_to_string_adaptive` uses Dragon4 utilities like `Dragon4_Positional_QuadDType` and `Dragon4_Scientific_QuadDType` for conversion and they both returns a `PyUnicode_FromString` object and from them we extract the cstring from it.

I can modify the dragon4 helper or add one more helper with `_cstr` suffix that returns cstring and then this would be doable. Let me know if this sounds good?